### PR TITLE
Enhance lora tests with more layer and rank variations

### PR DESCRIPTION
**Pull request by tterrysun**
Enhance lora tests with more layer and rank variations.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -8,6 +8,7 @@ transformers >= 4.38.0  # Required for Gemma.
 xformers == 0.0.23.post1  # Required for CUDA 12.1.
 fastapi
 uvicorn[standard]
+peft == 0.8.2
```
Comment (pcmoritz) [inline requirements.txt:11]: Given that this is only used for testing, it should go to `requirements-dev.txt`

File: tests/lora/test_layer_variation.py [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    "[system] Given a target sentence construct the underlying meaning representation\nof the input sentence as a single function with attributes and attribute\nvalues. This function should describe the target string accurately and the\nfunction must be one of the following ['inform', 'request', 'give_opinion',\n'confirm', 'verify_attribute', 'suggest', 'request_explanation',\n'recommend', 'request_attribute'].\n\nThe attributes must be one of the following:\n['name', 'exp_release_date', 'release_year', 'developer', 'esrb', 'rating',\n'genres', 'player_perspective', 'has_multiplayer', 'platforms',\n'available_on_steam', 'has_linux_release', 'has_mac_release', 'specifier'] [/system] [user] Here is the target sentence:\nBioShock is a good role-playing, action-adventure, shooter that released for PlayStation, Xbox, and PC in 2007. It is available on Steam, and it has a Mac release but not a Linux release. [/user] [assistant]",
+]
+TMP_PATH = "/mnt/local_storage/"
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:17]: remove this :)

File: tests/lora/conftest.py [hunk 3]
```diff
@@ -121,6 +124,14 @@ def sql_lora_files():
     return snapshot_download(repo_id="yard1/llama-2-7b-sql-lora-test")
 
 
+@pytest.fixture(scope="session")
```
Comment (pcmoritz) [inline tests/lora/conftest.py:127]: you don't need this fixture and `TMP_PATH` above. Pytest already has a built in `tmpdir` fixture that you can use https://docs.pytest.org/en/6.2.x/tmpdir.html#the-tmpdir-fixture

File: tests/lora/test_layer_variation.py [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+TARGET_MODULES_LIST = []
+for length in range(2, 6):
+    TARGET_MODULES_LIST.extend(
+        [sample(SUPPORTED_MODULES, length) for _ in range(3)])
+
+
+# Test the functionality when layer and rank are varied
+@pytest.mark.parametrize("target_modules", TARGET_MODULES_LIST)
+@pytest.mark.parametrize("rank", [8, 16, 32, 64])
+def test_layer_variation_functionality(target_modules, rank, tmp_path):
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:69]: We should just remove this test -- it is completely subsumed by `test_layer_variation_verify_reference`, right?

File: tests/lora/test_layer_variation.py [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+    # Print the outputs.
+    generated_texts = []
+    generated_logprobs = []
+    for output in outputs:
+        prompt = output.prompt
+        generated_text = output.outputs[0].text
+        generated_texts.append(generated_text)
+        print(f"Prompt: {prompt!r}, Generated text: {generated_text!r}")
+        generated_logprobs.append([
+            list(logprob.keys()) for logprob in outputs[0].outputs[0].logprobs
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:51]: Should this be `output.outputs[0].logprobs`? Otherwise you will only ever use the first prompt, right?

File: tests/lora/test_layer_variation.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    "[system] Given a target sentence construct the underlying meaning representation\nof the input sentence as a single function with attributes and attribute\nvalues. This function should describe the target string accurately and the\nfunction must be one of the following ['inform', 'request', 'give_opinion',\n'confirm', 'verify_attribute', 'suggest', 'request_explanation',\n'recommend', 'request_attribute'].\n\nThe attributes must be one of the following:\n['name', 'exp_release_date', 'release_year', 'developer', 'esrb', 'rating',\n'genres', 'player_perspective', 'has_multiplayer', 'platforms',\n'available_on_steam', 'has_linux_release', 'has_mac_release', 'specifier'] [/system] [user] Here is the target sentence:\nBioShock is a good role-playing, action-adventure, shooter that released for PlayStation, Xbox, and PC in 2007. It is available on Steam, and it has a Mac release but not a Linux release. [/user] [assistant]",
+]
+TMP_PATH = "/mnt/local_storage/"
+
+
+def get_lora_model(model_id: str, target_modules: List[str], rank: int):
```
[the remark this answers is not in the document]
Comment (tterrysun) [reply tests/lora/test_layer_variation.py:19]: it's a default initialized lora, we use the merged one as golden reference to verify the correctness, the lora weights won't matter as long as we're using the same one

File: tests/lora/test_layer_variation.py (hunk 5 above)
Comment (tterrysun) [reply tests/lora/test_layer_variation.py:50]: this is intentional, the following tokens are conditioned on the preceding ones and numerical difference could accumulate, the overall goal of this test is to be essential but not sufficient, i.e. if this test passes it doesn't mean much, if this fails something must be wrong. it's also the reason why we only compare the top-5 set without caring about the order

File: tests/lora/test_layer_variation.py (hunk 4 above)
Comment (tterrysun) [reply tests/lora/test_layer_variation.py]: removed!

File: tests/lora/test_layer_variation.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (pcmoritz) [reply tests/lora/test_layer_variation.py:19]: Can you point to where in the docs it says it is a default LoRA and what it is? That part was not clear to me (maybe add a comment)

File: tests/lora/test_layer_variation.py (hunk 5 above)
Comment (pcmoritz) [reply tests/lora/test_layer_variation.py:50]: Wait, but this is only ever using the first prompt (i.e. `PROMPT[0]`) if I understand this correctly -- that can't possibly be your intention. Otherwise why even include the other prompts?

Comment (tterrysun) [reply tests/lora/test_layer_variation.py]: yes you're right, I misread you comment. fixed

**Comment by pcmoritz** [comment]
On a high level, I think these tests would be much better understandable if it was a single test that evaluates the reference and then tests against the target, instead of having two tests and hardcoding the taget.

I also find it very irritating that we only test correctness of the first token for each sequence. Is there a way to test the rest of the sequence too and still have it be robust? Other tests do similar things, right? Maybe instead of testing the predicted tokens, you can test that the logits are numerically close.

File: tests/lora/conftest.py [hunk 7]
```diff
@@ -21,6 +21,8 @@
 from vllm.model_executor.parallel_utils.parallel_state import (
     destroy_model_parallel, initialize_model_parallel)
 
+TMP_PATH = "/mnt/local_storage/"
```
Comment (pcmoritz) [inline tests/lora/conftest.py:24]: This should be removed

File: tests/lora/test_layer_variation.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+                   max_loras=4,
+                   tensor_parallel_size=tp_size,
+                   worker_use_ray=True)
+    model = get_lora_model(MODEL_PATH, target_modules, rank)
+    tmp_dir_lora = os.path.join(tmpdir, "tmp_dir_lora")
+    model.save_pretrained(tmp_dir_lora)
+    merged_probs = do_sample(llm, tmp_dir_lora, 1, logprobs=5, n_tokens=32)
+    del llm
+    cleanup()
+    shutil.rmtree(str(tmpdir))
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:89]: If you need to delete the temp dir, it will be better to use
```
with tempfile.TemporaryDirectory(delete=True) as tmpdir:
    # Code that uses temp dir here
```

File: tests/lora/test_layer_variation.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+
+def do_sample(llm,
+              lora_path: Optional[str] = None,
+              lora_id: Optional[int] = None,
+              logprobs: int = 0,
+              n_tokens: int = 256):
+    prompts = PROMPTS
+    sampling_params = vllm.SamplingParams(temperature=0,
+                                          max_tokens=256,
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:36]: Is there a reason btw you are not setting `max_tokens=n_tokens` here and then skip the slicing below?

File: tests/lora/test_layer_variation.py [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+        tmp_dir_lora = os.path.join(tmpdir, "tmp_dir_lora")
+        model.save_pretrained(tmp_dir_lora)
+        merged_probs = do_sample(llm, tmp_dir_lora, 1, logprobs=5, n_tokens=32)
+    del llm
+    cleanup()
+    reference_id_sets = [set(prob[0]) for prob in merged_probs]
+
+    model = get_lora_model(MODEL_PATH, target_modules, rank)
+    with tempfile.TemporaryDirectory() as tmpdir:
+        tmp_dir_merged = os.path.join(tmpdir, "tmp_dir_merged")
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:94]: Is there a need to introduce an additional layer of tmp directories? Same above.

File: tests/lora/test_layer_variation.py [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        merged_model = model.merge_and_unload()
+        merged_model.save_pretrained(tmp_dir_merged)
+        llm = vllm.LLM(tmp_dir_merged,
+                       tokenizer=MODEL_PATH,
+                       enable_lora=False,
+                       max_num_seqs=16,
+                       tensor_parallel_size=4,
```
Comment (pcmoritz) [inline tests/lora/test_layer_variation.py:101]: This shouldn't be hard coded (this won't work in the CI)

**Review by Yard1 — APPROVED** [review]

**Review by pcmoritz — APPROVED** [review]
Nice! There are other tests where
```
    if torch.cuda.device_count() < tp_size:
```
is causing problems, it might be better to remove that?

**Comment by richardliaw** [comment]
can we merge?
